src/recon_buildblock/KappaComputation.cxx

```diff
@@ -0,0 +1,264 @@
+*/
+
+#include "stir/recon_buildblock/KappaComputation.h"
+#include "stir/info.h"
+#include "stir/modelling/ParametricDiscretisedDensity.h"
+#include "stir/IO/OutputFileFormat.h"
+#include "stir/IO/read_from_file.h"
+#include "stir/is_null_ptr.h"
+//#include "stir/recon_buildblock/GeneralisedObjectiveFunction.h"
+//#include "stir/recon_buildblock/GeneralisedPrior.h"
+#include "stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.h"
+#include "stir/unique_ptr.h"
+
+
+START_NAMESPACE_STIR
+
+template <typename TargetT>
+KappaComputation<TargetT>::
+KappaComputation()
+{
+  this->set_defaults();
+}
+
+template <typename TargetT>
+void
+KappaComputation<TargetT>::
+set_defaults()
+{
+  objective_function_sptr.reset(new PoissonLogLikelihoodWithLinearModelForMeanAndProjData<TargetT>);
+  output_file_format_sptr = OutputFileFormat<TargetT>::default_sptr();
+  set_use_approximate_hessian(true);
+  set_compute_with_penalty(false);
+}
+
+template <typename TargetT>
+void
+KappaComputation<TargetT>::initialise_keymap()
+{
+  parser.add_start_key("Kappa Computation Parameters");
+  parser.add_key("kappa filename", &kappa_filename);
+  parser.add_key("input image filename", &input_image_filename);
+  parser.add_key("use approximate hessian", &use_approximate_hessian);
+  parser.add_key("compute with penalty", &compute_with_penalty);
+  parser.add_parsing_key("objective function type", &objective_function_sptr);
+  parser.add_stop_key("End");
+}
+
+template <typename TargetT>
+bool
+KappaComputation<TargetT>::post_processing()
+{
+  if (is_null_ptr(this->objective_function_sptr))
+  {
+    error("objective_function_sptr is null");
+    return true;
+  }
+
+  if (input_image_filename.empty())
+  {
+    error("Please define input_image_filename.");
+    return true;
+  }
+  return false;
+}
+
+template <typename TargetT>
+PoissonLogLikelihoodWithLinearModelForMean<TargetT > const&
+KappaComputation<TargetT>::
+get_objective_function()
+{
+  return static_cast<PoissonLogLikelihoodWithLinearModelForMean<TargetT >&> (*objective_function_sptr);
+}
+
+template <typename TargetT>
+void
+KappaComputation<TargetT>::
+set_objective_function_sptr(const shared_ptr<PoissonLogLikelihoodWithLinearModelForMean<TargetT > >& obj_fun)
+{
+  this->objective_function_sptr  = obj_fun;
+}
+
+template <typename TargetT>
+shared_ptr <TargetT>
+KappaComputation<TargetT>::
+get_input_image()
+{
+  return input_image;
+}
+
+template <typename TargetT>
+void
+KappaComputation<TargetT>::
+set_input_image(shared_ptr <TargetT > const& image)
+{
+  input_image = image;
+}
+
+template <typename TargetT>
+shared_ptr <TargetT>
+KappaComputation<TargetT>::
+get_kappa_image_target_sptr()
+{
+  return kappa_image_target_sptr;
+}
+
+template <typename TargetT>
+void
+KappaComputation<TargetT>::
+reset_kappa_image_target_sptr(shared_ptr <TargetT > const& image)
+{
+  kappa_image_target_sptr =  unique_ptr<TargetT>(image->get_empty_copy());
+  kappa_image_target_sptr->fill(0);
```

Comment:
Does `get_empty_copy()` already `fill(0)`? If so, L136 is not needed

Reply:
probably does, but not guaranteed, so best to fill anyway